### PR TITLE
Remove old resources

### PR DESCRIPTION
**Pull request by eliecharra**

### CODE REVIEW

File: pkg/resource/resource.go [hunk 1]
```diff
@@ -15,83 +15,19 @@ import (
...
-	return false
+	Attributes() *Attributes
+	Schema() *Schema
 }
 
 type AbstractResource struct {
 	Id    string
 	Type  string
 	Attrs *Attributes
+	Sch   *Schema `json:"-" diff:"-"`
```
Comment (eliecharra) [inline pkg/resource/resource.go:26]: Why `diff:"-"` ? 

BTW I really have responsibility issue with this. I don't get why we do not keep resource and schema separated. For me there is not need to inject the schema in the resource they are two distinct struct (like cty value and cty type)

We may have a vocal talk about it

File: pkg/remote/aws/test/iam_role_policy_attachment_multiple/.terraform.lock.hcl [hunk 2]
```diff
@@ -0,0 +1,20 @@
+# This file is maintained automatically by "terraform init".
```
Comment (eliecharra) [inline pkg/remote/aws/test/iam_role_policy_attachment_multiple/.terraform.lock.hcl:1]: This file is useless

File: pkg/remote/aws/test/iam_user_policy_attachment_multiple/.terraform.lock.hcl [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# This file is maintained automatically by "terraform init".
```
Comment (eliecharra) [inline pkg/remote/aws/test/iam_user_policy_attachment_multiple/.terraform.lock.hcl:1]: This file is useless

File: pkg/resource/deserializer.go [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+func NewDeserializer(factory ResourceFactory) *Deserializer {
+	return &Deserializer{factory}
+}
+
+func (s *Deserializer) Deserialize(ty string, rawList []cty.Value) ([]Resource, error) {
+	resources := make([]Resource, 0)
+	for _, rawResource := range rawList {
+		rawResource := rawResource
+		attrs := ToResourceAttributes(&rawResource)
```
Comment (eliecharra) [inline pkg/resource/deserializer.go:19]: What's the purpose of `ToResourceAttributes` ? This logic do not need to be in a dedicated exported method, it does a deserialization of cty values object and it is exactly the purpose of this deserializer.

File: pkg/resource/aws/testdata/acc/aws_security_group/terraform.tf [hunk 5]
```diff
@@ -72,26 +67,6 @@ resource "aws_security_group" "sg-bis-tutu" {
     }
 }
 
-resource "aws_security_group_rule" "tutu-egress" {
```
Comment (eliecharra) [inline pkg/resource/aws/testdata/acc/aws_security_group/terraform.tf:75]: I don't get why do we have diff on theses acceptance test here

File: pkg/resource/aws/testdata/acc/aws_iam_access_key/terraform.tf [hunk 6]
```diff
@@ -16,3 +16,13 @@ resource "aws_iam_user" "testuser_access_key" {
 resource "aws_iam_access_key" "accesskey" {
     user = aws_iam_user.testuser_access_key.name
 }
+
+resource "aws_iam_access_key" "ak" {
+    user    = aws_iam_user.test.name
+    pgp_key = "keybase:test"
```
Comment (eliecharra) [inline pkg/resource/aws/testdata/acc/aws_iam_access_key/terraform.tf:22]: Is this related to this PR ? 🤔

File: pkg/resource/resource.go [hunk 7]
```diff
@@ -102,12 +38,20 @@ func (a *AbstractResource) TerraformType() string {
 	return a.Type
 }
 
-func (a *AbstractResource) CtyValue() *cty.Value {
-	return nil
+func (a *AbstractResource) Attributes() *Attributes {
+	return a.Attrs
+}
+
+func (a *AbstractResource) HumanReadableAttributes() map[string]string {
```
Comment (eliecharra) [inline pkg/resource/resource.go:45]: I think this logic should not be related to the resource struct.
 
We put too many thinks in resource : the schema, the data itself, the human readable function that is bound to the schema 🤔 not very clear IMHO we should rethink this.

File: pkg/resource/resource.go [hunk 8]
```diff
@@ -249,6 +201,26 @@ func (a *Attributes) GetFloat64(path string) *float64 {
 	return &v
 }
 
+func (a *Attributes) GetMap(path string) map[string]interface{} {
```
Comment (eliecharra) [inline pkg/resource/resource.go:204]: This should be unexported or done directly in `GetStringMap`.

BTW make a full copy of the map looks heavy for a simple type cast and I don't get why we don't do this :

image

File: pkg/resource/resource.go [hunk 9]
```diff
@@ -215,11 +163,15 @@ func (a *Attributes) GetString(path string) *string {
...
 
 func (a *Attributes) GetStringSlice(path string) []string {
-	val, exist := (*a)[path]
-	if !exist {
+	val := a.GetSlice(path)
+	if val == nil {
 		return nil
 	}
-	return val.([]string)
+	slice := make([]string, 0, len(val))
```
Comment (eliecharra) [inline pkg/resource/resource.go:170]: Why do we need to make a copy ?

File: pkg/resource/resource.go [hunk 10]
```diff
@@ -15,83 +15,19 @@ import (
...
-}
-
-func IsRefactoredResource(typ string) bool {
-	for _, refactoredResource := range refactoredResources {
-		if typ == refactoredResource {
-			return true
-		}
-	}
-	return false
+	Attributes() *Attributes
```
Comment (eliecharra) [inline pkg/resource/resource.go:18]: I had planned to remove the resource interface and rename `AbstractResource` to `Resource` WDYT ? I don't think this interface makes sense anymore

File: test/resource/resource.go [hunk 11]
```diff
@@ -4,50 +4,48 @@ import (
 	"github.com/cloudskiff/driftctl/pkg/resource"
 	"github.com/cloudskiff/driftctl/test/schemas"
 	"github.com/hashicorp/terraform/providers"
-	"github.com/zclconf/go-cty/cty"
 )
 
 type FakeResource struct {
```
[the remark this answers is not in the document]
Comment (moadibfr) [reply test/resource/resource.go:9]: Yeah we asked ourselves the same question. I think we could remove this (and maybe remove resource interface and only use abstractresource named as resource). But this will mean re writing all the test that use it and this is not easy for every test :/

[the remark this answers is not in the document]
Comment (wbeuil) [reply test/resource/resource.go:9]: We could totally ditch the `FakeResource`

File: pkg/resource/aws/testdata/acc/aws_security_group/terraform.tf (hunk 5 above)
Comment (wbeuil) [reply pkg/resource/aws/testdata/acc/aws_security_group/terraform.tf:75]: I remove those security group rules since they are useless in this acceptance test. They were copy/pasted in the `security_group_rule` acc tests.

File: pkg/resource/resource.go (hunk 1 above)
Comment (moadibfr) [reply pkg/resource/resource.go:26]: We diff AbstractResource entirely in some tests so we need to ignore schema.
I quite like that the schema is given with the concerned resource instead of having to inject a separate repository to get it. I thought there is no responsibility concern here as the resource and the schema are directly linked and we only build the resource in one place. But yeah let's talk about this tomorrow. (yet I have to add that cty value have their own type embedded :grimacing:  https://github.com/zclconf/go-cty/blob/main/cty/value.go )

File: pkg/resource/aws/testdata/acc/aws_iam_access_key/terraform.tf (hunk 6 above)
Comment (wbeuil) [reply pkg/resource/aws/testdata/acc/aws_iam_access_key/terraform.tf:22]: That's what @sjourdan told us in the last vocal meeting `if you can put it in the refacto, go`. But I can revert the last commit.

File: pkg/remote/aws/test/iam_role_policy_attachment_multiple/.terraform.lock.hcl (hunk 2 above)
Comment (moadibfr) [reply pkg/remote/aws/test/iam_role_policy_attachment_multiple/.terraform.lock.hcl:1]: should we add version in the associated tf to lock something or just delete it ?

File: pkg/resource/deserializer.go (hunk 4 above)
Comment (moadibfr) [reply pkg/resource/deserializer.go:19]: yeah you're right. I think this is a legacy method that we did not move

File: pkg/resource/resource.go (hunk 8 above)
Comment (wbeuil) [reply pkg/resource/resource.go:204]: I added both the `GetMap` and `GetMapString` methods to have more method utilities in our pocket. And it totally make sense to only have the GetMap and cast the value at the end, didn't think of it.

File: pkg/resource/resource.go (hunk 7 above)
Comment (moadibfr) [reply pkg/resource/resource.go:45]: yeah I think this is the same discussion as for separation between the schema and the resource. It is handy that abstractresource just implement the HumanReadableAttributes interface I think

File: pkg/resource/resource.go (hunk 9 above)
Comment (wbeuil) [reply pkg/resource/resource.go:170]: Just like the `GetMap` we can't use the `[]string` cast on `[]interface{}`. I don't remember the resource I got the error ..

File: pkg/resource/resource.go (hunk 10 above)
Comment (wbeuil) [reply pkg/resource/resource.go:18]: I agree !

File: pkg/resource/resource.go (hunk 8 above)
Comment (moadibfr) [reply pkg/resource/resource.go:204]: this was supposed to be one of the getter/autocast function like GetSlice, GetString, ...
And yeah I think we could replace the getstringmap you're right

File: pkg/remote/aws/test/iam_role_policy_attachment_multiple/.terraform.lock.hcl (hunk 2 above)
Comment (wbeuil) [reply pkg/remote/aws/test/iam_role_policy_attachment_multiple/.terraform.lock.hcl:1]: Just delete it

File: pkg/resource/resource.go (hunk 9 above)
Comment (moadibfr) [reply pkg/resource/resource.go:170]: same as getStringmap it is not possible to cast []inteferace{} to []string

File: pkg/resource/resource.go (hunk 10 above)
Comment (moadibfr) [reply pkg/resource/resource.go:18]: yeah as I said in a response before this one I think it would be great but rewriting all the tests is really painfull :cry:

File: pkg/resource/resource.go (hunk 7 above)
Comment (eliecharra) [reply pkg/resource/resource.go:45]: For me this is more related to the `View` than the `Model` part if we think MVC. I had in mind that the resource is our model and is only a data container object. This should not be bound to a resource.

File: test/resource/resource.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (eliecharra) [reply test/resource/resource.go:9]: We'll do this in another PR

**Review by moadibfr — APPROVED** [review]
